065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,298 @@
+
+### Error handling
+
+In case of error, the replicas change will remain pending or ongoing (no status update), and will be indefinitely retried by the periodic reconciliation.
+All errors will be written in the Topic Operator logs and in the KafkaTopic status.
+
+In Cruise Control, there is no way to differentiate between temporary errors (resolve by themself) from permanent errors (require some action).
+For example, the `NotEnoughValidWindowsException` can be raised when Cruise Control has just started and the cluster model is still building, but also when there is a configuration error which prevents broker metrics collection.
+A similar example is the `OptimizationFailureException` raised when some hard goal is violated, which can be temporary if the configured network capacity is less than the real network capacity and there is a traffic peak.
+Instead, the `OngoingExecutionException` is clearly temporary, and it means that a task execution was attempted while there is already one that is being executed by Cruise Control.
+Another temporary example is the `RuntimeException` when Cruise Control's task queue is full.
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause disruption to producers with `acks=all`.
+When this happens, the Topic Operator won't block the operation, but will just log a warning, because the KafkaRoller ignores topics with RF < minISR, and they don't even show up as under replicated in Kafka metrics.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+When a managed KafkaTopic with replicas change is deleted in Kubernetes, the Topic Operator will also delete the topic in Kafka, but the task execution will continue in Cruise Control.
+There is little benefit in using the `stop_proposal_execution` endpoint, because most replicas change tasks have only one batch that can't be stopped.
+No task remains hanging, the only problem is that Cruise Control does some unnecessary work executing a batch for a topic that has been deleted.
+
+When a managed topic with replicas change is deleted in Kafka, the Topic Operator will recreate the topic with the target replication factor.
+
+When Topic Operator is restarted, both pending and ongoing replicas changes will be recovered from the KafkaTopic status.
+The ongoing state will include a `sessionId`, which is the unique identifier of a Cruise Control task, which includes one or more replicas changes.
+
+When Cruise Control is restarted, all active tasks are lost because there is no persistent memory.
+The Topic Operator will detect this event when the ongoing changes check returns no tasks for a known `sessionId`.
+All pending and ongoing replicas changes will still be stored in KafkaTopic status, and new tasks will be created as soon as Cruise Control becomes ready.
+
```

Comment:
Should we update the CRD with an `AdditionalPrinterColumn` for the `status.replicasChange.state`?

Reply:
Given it's a temporary state, I'm not sure we need this.